sos/collector/clusters/juju.py

```diff
@@ -0,0 +1,174 @@
+# Copyright (c) 2023 Canonical Ltd., Chi Wai Chan <chiwai.chan@canonical.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import json
+import collections
+import re
+from typing import Optional
+from sos.collector.clusters import Cluster
+
+
+def _parse_option_string(strings: Optional[str] = None):
+    """Parse commad separated string."""
+    if not strings:
+        return []
+    return [string.strip() for string in strings.split(",")]
+
+
+class juju(Cluster):
+    """
+    The juju cluster profile is intended to be used on juju managed clouds.
+    It"s assumed that `juju` is installed on the machine where `sos` is called,
+    and that the juju user has superuser privilege to the current controller.
+
+    By default, the sos reports will be collected from all the applications in
+    the current model. If necessary, you can filter the nodes by models /
+    applications / units / machines with cluster options.
+
+    Example:
+
+    sos collect --cluster-type juju -c "juju.models=sos" -c "juju.apps=a,b,c"
+
+    """
+
+    cmd = "juju"
+    cluster_name = "Juju Managed Clouds"
+    option_list = [
+        ("apps", "", "Filter node list by apps (comma separated regex)."),
+        ("units", "", "Filter node list by units (comma separated string)."),
+        ("models", "", "Filter node list by models (comma separated string)."),
+        (
+            "machines",
+            "",
+            "Filter node list by machines (comma separated string).",
+        ),
+    ]
+
+    def _cleanup_juju_output(self, output):
+        """Remove leading characters before {."""
+        return re.sub(r"(^[^{]*)(.*)", "\\2", output, 0, re.MULTILINE)
+
+    def _get_model_info(self, model_name):
+        juju_status = self._execute_juju_status(model_name)
+
+        index = collections.defaultdict(dict)
+        self._add_principals_to_index(index, juju_status, model_name)
+        self._add_subordinates_to_index(index, juju_status, model_name)
+        self._add_machines_to_index(index, juju_status, model_name)
+
+        return index
```

Comment:
It might be beneficial to have a comment explaining purpose of the method, optionally a short example of "juju status" output and expected `index` dictionary - or at least a pointer to `tests/unittests/juju/data/juju_output_sos.json` :)

Reply:
Refactored this part